cabal-install/Distribution/Client/ProjectOrchestration.hs

```diff
@@ -243,14 +244,33 @@ runProjectBuildPhase verbosity ProjectBuildContext {..} =
 -- | Adjust an 'ElaboratedInstallPlan' by selecting just those parts of it
 -- required to build the given user targets.
 --
--- How to get the 'PackageTarget's from the 'UserBuildTarget' is customisable.
+-- How to get the 'PackageTarget's from the 'UserBuildTarget' is customisable,
+-- so that we can change the meaning of @pkgname@ to target a build or
+-- repl depending on which command is calling it.
 --
-selectTargets :: PackageTarget
+-- Conceptually, every target identifies one or more roots in the
+-- 'ElaboratedInstallPlan', which we then use to determine the closure
+-- of what packages need to be built, dropping everything from
+-- 'ElaboratedInstallPlan' that is unnecessary.
+--
+-- There are a few complications, however:
+--
+--      * Strangely (in ezyang's opinion) an ElaboratedConfiguredPackage
+--        doesn't actually dictate what should be built: we're asked
+--        to fill in 'pkgBuildTargets', 'pkgReplTarget' and
+--        'pkgBuildHaddock' to control what we actually do in the
+--        build phase.  This means we have to fiddle around with
+--        the ElaboratedConfiguredPackage roots to say what it will
+--        build.
```

Comment:
I don't think this criticism is very clear. It reads like it assumes an alternative is obvious to the reader, but it's not obvious to me.